fendermint/vm/interpreter/src/fvm/checkpoint.rs

```diff
@@ -212,3 +241,164 @@ where
 
     Ok(PowerTable(power_table))
 }
+
+/// Convert the collaterals and metadata in the membership to the public key and power expected by the system.
+fn membership_to_power_table(m: &Membership, power_scale: PowerScale) -> PowerTable {
+    let mut pt = Vec::new();
+
+    for v in m.validators.iter() {
+        // Ignoring any metadata that isn't a public key.
+        if let Ok(pk) = PublicKey::parse_slice(&v.metadata, None) {
+            let c = u256_to_tokens(v.weight);
+            pt.push(Validator {
+                public_key: ValidatorKey(pk),
+                power: Collateral(c).into_power(power_scale),
+            })
+        }
+    }
+
+    PowerTable(pt)
+}
+
+fn u256_to_tokens(value: U256) -> TokenAmount {
+    let mut bz = [0u8; 32];
+    value.to_big_endian(&mut bz);
+    let atto = BigInt::from_bytes_be(Sign::Plus, &bz);
+    TokenAmount::from_atto(atto)
+}
+
+/// Calculate the difference between the current and the next power table, to return to CometBFT only what changed.
+fn power_diff(current: &PowerTable, next: PowerTable) -> PowerUpdates {
+    let current = current
```

Comment:
Maybe there's a type mismatch that I am missing, but why can't you use `into_power_map` here?

Reply:
Thanks for asking, the only reason is that `current` is a reference, while `into_power_map` consumes the input. I actually wanted to check if there is a way to unify these two because they are otherwise doing the same thing. 

`current` is a reference because:
* it's used later when constructing the Merkle tree, although maybe the tree could be constructed first
* I thought the values in it would not be used most of the time: the public key is only cloned if a validator is deleted

`next` is taken by value because:
* the value for new validators is going to be pushed into the output as-is
* it's not used for anything else

But I will have a second look because this is certainly jarring.

Reply:
Decided to rearrange the code so both arguments are consumed. Originally I also returned the power table from `maybe_create_checkpoint` because it was used to construct the proof, but since we re-send the validator signature for every potentially pending checkpoint, I just re-fetch it and ignored the returned value to keep the code simpler. Not a big deal.